File: keepr/Services/VaultsService.cs
namespace keepr.Services;

public class VaultsService
{
  private readonly VaultsRepository _vaultsRepository;

  public VaultsService(VaultsRepository vaultsRepository)
  {
    _vaultsRepository = vaultsRepository;
  }

  internal Vault CreateVault(Vault newVault)
  {
    return _vaultsRepository.Create(newVault);
  }

  internal Vault GetVaultById(int vaultId, string accountId)
  {
    Vault vault = _vaultsRepository.GetById(vaultId);
    if (vault == null)
    {
      throw new Exception("Vault does not exist");
    }
    if (vault.IsPrivate && vault.CreatorId != accountId)
    {
      throw new Exception("This vault is private");
    }
    return vault;
  }

  internal Vault GetVaultByIdNoAuth(int vaultId)
  {
    Vault vault = _vaultsRepository.GetById(vaultId);
    if (vault == null)
    {
      throw new Exception("Vault does not exist");
    }
    if (vault.IsPrivate)
    {
      throw new Exception("This vault is private");
    }
    return vault;
  }

  internal Vault EditVault(Vault vaultData, string accountId)
  {
    Vault orig = GetVaultById(vaultData.Id, accountId);
    if (orig.CreatorId != accountId)
    {
      throw new Exception("This vault is not yours, you cannot edit it");
    }
    orig.Name = vaultData.Name ?? orig.Name;
    orig.Description = vaultData.Description ?? orig.Description;
    orig.Img = vaultData.Img ?? orig.Img;
    orig.IsPrivate = vaultData.IsPrivate == false ? orig.IsPrivate : vaultData.IsPrivate;
    return _vaultsRepository.Update(orig);
  }

  internal void RemoveVault(int vaultId, string accountId)
  {
    Vault vault = GetVaultById(vaultId, accountId);
    if (vault.CreatorId != accountId)
    {
      throw new Exception("This vault is not yours, you cannot delete it");
    }
    _vaultsRepository.Delete(vault);
  }
}

File: keepr.client/src/services/VaultsService.js
import { AppState } from "../AppState.js"
import { Vault } from "../models/Vault.js"
import { router } from "../router.js"
import Pop from "../utils/Pop.js"
import { accountService } from "./AccountService.js"
import { api } from "./AxiosService.js"
import { vaultKeepsService } from "./VaultKeepsService.js"

class VaultsService {
  async createVault(vaultData) {
    const res = await api.post("/api/vaults", vaultData)
    const vault = new Vault(res.data)
    AppState.vaults.push(vault)
  }

  async getActiveVault(vaultId) {
    const vault = await this.getVaultById(vaultId)
    if (vault.isPrivate == true && vault.creatorId != AppState.account.id) {

      Pop.toast("That vault is private, sorry!", "warning")
      return
    }
    if (!vault) {
      Pop.toast("Unable to find vault, sorry!", "warning")
      return
    }
    await vaultKeepsService.getVaultKeeps(vaultId)
    AppState.activeVault = vault
  }

  async getVaultById(vaultId) {
    const res = await api.get(`/api/vaults/${vaultId}`)
    const vault = new Vault(res.data)
    if (!vault) {
      console.error("Could not find vault")
    }
    return vault
  }

  async removeVault(vaultId) {
    const vault = await this.getVaultById(vaultId)
    if (vault.creatorId != AppState.account.id) {
      Pop.error("You can only delete your own vaults")
    }
    await api.delete(`/api/vaults/${vaultId}`)
    router.push({ name: "Home" })
  }

  async editVault(vaultData) {
    const vault = await this.getVaultById(AppState.activeVault.id)
    if (vault.creatorId != AppState.account.id) {
      Pop.error("You can only edit your own vaults")
    }
    const res = await api.put(`/api/vaults/${vault.id}`, vaultData)
    const newVault = new Vault(res.data)
    AppState.activeVault = newVault
    accountService.getMyVaults()
  }
}

export const vaultsService = new VaultsService()